API Spec change for DNC RP

tamilmani1989 opened the pull request:
<i>MSFT employees can try out our new experience at <b>[OpenAPI Hub](https://aka.ms/openapiportal) </b> - one location for using our validation tools and finding your workflow. 
</i>

### Contribution checklist:
- [x] I commit to follow the [Breaking Change Policy](http://aka.ms/bcforapi) of “no breaking changes
- [x] I have reviewed the [documentation](https://aka.ms/ameonboard) for the workflow.
- [x] [Validation tools](https://aka.ms/swaggertools) were run on swagger spec(s) and errors have all been fixed in this PR. [How to fix?](https://aka.ms/ci-fix)

If any further question about AME onboarding or validation tools, please view the [FAQ](https://aka.ms/faqinprreview).

### ARM API Review Checklist
- [x] Ensure to check this box if one of the following scenarios meet updates in the PR, so that label “WaitForARMFeedback” will be added automatically to involve ARM API Review. Failure to comply may result in delays for manifest application. Note this does not apply to data plane APIs, all “removals” and “adding a new property” no more require ARM API review.
  - Adding new API(s)
  - Adding a new API version
  - Adding a new service

- [ ] If you are blocked on ARM review and want to get the PR merged with urgency, please get the ARM oncall for reviews (*RP Manifest Approvers* team under <ins>Azure Resource Manager service</ins>) from IcM and reach out to them. 

### Breaking Change Review Checklist 
If there are following updates in the PR, ensure to request an approval from API Review Board as defined in the [Breaking Change Policy](http://aka.ms/bcforapi). 

- [ ] Removing API(s) in stable version
- [ ] Removing properties in stable version
- [ ] Removing API version(s) in stable version
- [ ] Updating API in stable version with Breaking Change Validation errors
- [ ] Updating API(s) in preview over 1 year

Please follow the link to find more details on [PR review process](https://aka.ms/SwaggerPRReview).

tamilmani1989 committed:
updated apispec

tamilmani1989 committed:
spell error fix

tamilmani1989 committed:
removed dncappid from response

tamilmani1989 committed:
updated apispec

tamilmani1989 committed:
Added appid and endpoint in response

tamilmani1989 committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs

tamilmani1989 committed:
adde controller id in delegated subnet response

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 committed:
updated examples

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 committed:
updated

AutorestCI commented:
/azp run automation - sdk

lirenhe commented:
@tamilmani1989 , it looks the referenced file specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/examples/delegatedNetworkControllerListByRG.json does not exist that caused a couple of tool check errors.

tamilmani1989 committed:
updated apispec

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 committed:
spell check and prettier fix

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 committed:
spell check fix

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 committed:
updated title same in all files

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 committed:
updated operationid

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 commented:
@lirenhe fixed and everything passed..can you please approve?

lirenhe commented:
> @lirenhe fixed and everything passed..can you please approve?

Could you also resolve the conflictg?

tamilmani1989 committed:
Merge branch 'master' of https://github.com/Azure/azure-rest-api-specs into delsubnet_resp

# Conflicts:
#	specification/dnc/resource-manager/readme.md

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 committed:
reverting to new change

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 commented:
> > @lirenhe fixed and everything passed..can you please approve?
> 
> Could you also resolve the conflict? And could you also highlight the change in your PR?

resolved the conflict

Main changes:

- added new resource orchestratorInstaces
- changed delegatedSubnets from extension resource to tracked resource
- added few fields in apischema

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/delegatedSubnets.json:
```diff
@@ -0,0 +1,515 @@
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DelegatedNetwork/delegatedSubnets/{resourceName}": {
```
changing from an extension/proxy resource to tracked is a massive change for consumers.  How are you going to backfill the existing subnets into the ARM tracked resource cache? Are you breaking your existing customers here?

tamilmani1989 replied:
no customer using our RP yet. Also removed all resources created, so there is no impact

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/delegatedSubnets.json:
```diff
@@ -0,0 +1,515 @@
+        "x-ms-long-running-operation": true
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.DelegatedNetwork/delegatedSubnetInstances": {
```
why is this referred to as delegatedSubnets in some places and delegatedSubnetInstances in others?

tamilmani1989 replied:
delegatedSubnets is resource and delegatedSubnetInstances list of delegatedSubnets resources under the subscription..just to differentiate it, i gave different name

pilor replied:
The collection GET at parent scopes should use the same resourceType name.  It is still returning a list of delegatedSubnets

tamilmani1989 replied:
sure i will update it

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/delegatedSubnets.json:
```diff
@@ -0,0 +1,515 @@
+          "readOnly": true,
+          "description": "Resource guid."
+        },
+        "state": {
```
long running PUTs should be using a "provisioningState" property, not "state"

tamilmani1989 replied:
state has "provisioning" as value.

pilor replied:
Yes.  The async pattern described here is strict:  https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/Addendum.md#creatingupdating-using-put

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/delegatedSubnets.json:
```diff
@@ -0,0 +1,515 @@
+          "description": "orchestrator details",
+          "$ref": "#/definitions/subnetDetails"
+        },
+        "controllerDetails": {
```
If subnets belong to a controller why not have them as children of the controller?  Can they switch controllers during their lifetime?  Do they have to be the same region as the subnet/controller?  Who validates that?

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/orchestratorInstances.json:
```diff
@@ -0,0 +1,540 @@
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DelegatedNetwork/orchestratorInstance/{resourceName}": {
```
resource types should be plural.  "orchestratorInstances".  Also, why use the "instance" suffix?

tamilmani1989 replied:
same reason explained above

tamilmani1989 replied:
the reason for instance is there can be multiple instances but the put one resource corresponds to one instance

pilor replied:
Thats true for all resource types.  There can be multiple "Microsoft.Compute/virtualMachines" but the PUT of a virtualMachines resource type is one VM.

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/controller.json:
```diff
@@ -0,0 +1,512 @@
+          "readOnly": true,
+          "description": "Resource guid."
+        },
+        "state": {
```
"provisioningState" for async PUT

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/orchestratorInstances.json:
```diff
@@ -0,0 +1,540 @@
+          "readOnly": true,
+          "description": "Resource guid."
+        },
+        "state": {
```
provisioningState

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/orchestratorInstances.json:
```diff
@@ -0,0 +1,540 @@
+          "description": "AAD ID used with apiserver",
+          "type": "string"
+        },
+        "properties": {
```
why are the properties of the resource buried 3 levels deep?

tamilmani1989 replied:
properties are selected based on orhcestratorType. thats y

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/orchestratorInstances.json:
```diff
@@ -0,0 +1,540 @@
+      "description": "Properties of orchestrator",
+      "type": "object",
+      "properties": {
+        "orchestratorType": {
```
what does this description mean?  The name of the property is "orchestratorType" but the description talks about AAD IDs.  If the values are restriced can this be a string enum?

tamilmani1989 replied:
will fix it

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/orchestratorInstances.json:
```diff
@@ -0,0 +1,540 @@
+          "description": "TenantID of server App ID",
+          "type": "string"
+        },
+        "clusterRootCA": {
```
in what format?  Is this a secret?

tamilmani1989 replied:
this is public root CA certificate..not secret

pilor replied:
In what format?  The description should make it clear what is expected here.  I.e. is it a base64 encoded public key, a name, etc...

tamilmani1989 replied:
sure will update it..its base64 encoded

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/orchestratorInstances.json:
```diff
@@ -0,0 +1,540 @@
+          "description": "RootCA certificate of kubernetes cluster",
+          "type": "string"
+        },
+        "apiServerEndpoint": {
```
the user provides this?

tamilmani1989 replied:
yes k8s apiServer url

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/orchestratorInstances.json:
```diff
@@ -0,0 +1,540 @@
+      "description": "controller details",
+      "type": "object",
+      "properties": {
+        "orchestratorAppID": {
```
What will the app ID be used for?  This can't be a managed identity?

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/orchestratorInstances.json:
```diff
@@ -0,0 +1,540 @@
+          "description": "orchestrator details",
+          "$ref": "#/definitions/OrchestratorInstanceDetails"
+        },
+        "controllerDetails": {
```
same question as subnets, can this be a child of a controller if it is associated with one?

tamilmani1989 committed:
addressed review comments

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 committed:
addressed feedback comments

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 committed:
spell check fix

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 committed:
updated readme.md

AutorestCI commented:
/azp run automation - sdk

tamilmani1989 committed:
updated definition name

AutorestCI commented:
/azp run automation - sdk

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/controller.json:
```diff
@@ -0,0 +1,512 @@
+              "$ref": "#/definitions/DelegatedController"
+            }
+          },
+          "201": {
```
patch should only return 200

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/controller.json:
```diff
@@ -0,0 +1,512 @@
+    }
+  },
+  "paths": {
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DelegatedNetwork/controller/{resourceName}": {
```
the resource type name needs to be plural ("controllers")

tamilmani1989 replied:
can't update this..since updated nrp code with same name and its in production and also some resources created with this type

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/controller.json:
```diff
@@ -0,0 +1,512 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.DelegatedNetwork/delegatedControllers": {
```
this should just be "controllers" not "delegatedControllers".  Its returning all the controllers in the subscription

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/controller.json:
```diff
@@ -0,0 +1,512 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.DelegatedNetwork/delegatedControllers": {
```
"controllers", not "delegatedControllers"

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/controller.json:
```diff
@@ -0,0 +1,512 @@
+          "readOnly": true,
+          "description": "The current state of dnc controller resource."
+        },
+        "controllerProperties": {
```
What is the first party app ID being used for?  This was supposed to transition to using managed identity

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/orchestrators.json:
```diff
@@ -0,0 +1,562 @@
+              "$ref": "#/definitions/Orchestrator"
+            }
+          },
+          "201": {
```
patch should only return 200

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/orchestrators.json:
```diff
@@ -0,0 +1,562 @@
+          "description": "K8s APIServer url",
+          "type": "string"
+        },
+        "controllerDetails": {
```
If the managed identity is associated with the orchestrator what is the dncAppId exposed in the controller still used for?  Can it be removed from that resource?

tamilmani1989 replied:
explained above

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/orchestrators.json:
```diff
@@ -0,0 +1,562 @@
+          "description": "controller details",
+          "$ref": "#/definitions/ControllerDetails"
+        },
+        "identity": {
```
identity should be a top level property on the resource (next to name, id, type)

tamilmani1989 replied:
thanks for pointing this out

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/delegatedSubnets.json:
```diff
@@ -0,0 +1,515 @@
+              "$ref": "#/definitions/DelegatedSubnet"
+            }
+          },
+          "201": {
```
patch should only return 200

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/delegatedSubnets.json:
```diff
@@ -0,0 +1,515 @@
+        "x-ms-long-running-operation": true
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.DelegatedNetwork/delegatedSubnets": {
```
Your subscription and resource group level GET collection calls need to be marked as x-ms-pageable and include "nextLink" in their response contract.  Even if your RP isn't returning nextLinks ARM will add a nextLink if the fanout to all your regions partially fails

pilor replied:
This applies to all 3 resource types in this PR

tamilmani1989 committed:
addressed feedback comments

AutorestCI commented:
/azp run automation - sdk

pilor commented on specification/dnc/resource-manager/Microsoft.DelegatedNetwork/preview/2020-08-08-preview/controller.json:
```diff
@@ -0,0 +1,501 @@
+      "description": "An array of Delegated controller resources.",
+      "type": "object",
+      "properties": {
+        "value": {
```
All 3 collection responses also need the "nextLink" property at the same level as "value".

tamilmani1989 replied:
yes updated it. Thanks

tamilmani1989 committed:
fixed lint issue

AutorestCI commented:
/azp run automation - sdk
